### DISCUSSION TITLE
tmp.imp.dir and cleanup

### DISCUSSION THREAD

**Discussion by SimpleAddress4390** [dcomtois/summarytools#191] [Ideas]
Hi.  I've been playing around with dfSummary and rMarkdown for a few months now.  I have read conflicting posts about how to generate the nice graphic images.
![ds0002](https://github.com/dcomtois/summarytools/assets/7174635/1583bc20-ca65-41a9-9c33-f8f4f8bb028b)

Even some of those posts suggest     method='render'   obviates the need for a tmp.img.dir directory in st_options, I've never gotten it to work.  So, I slog on, using a /tmp folder that fills up with thousands of png files. 

Has anyone found a workaround to this?  Or better, yet can dfSummary just wrote those pngs to the place that the knit/render process is already storing them to create the final html file?  I intercepted the location at the very end of the render, and held open a folder with all the same pngs there, the folder normally gets removed after the html is created.  It must be copying the pngs from the /tmp folder, so... why not have the \tmp folder be INSIDE the ...\local\temp\.....\ folder to begin with?  No user cleanup needed, no need to specify, etc.

Maybe I'm missing something (sure I am).

"C:/Program Files/RStudio/resources/app/bin/quarto/bin/tools/pandoc" +RTS -K512m -RTS Step01-Prep-1.01.knit.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output pandoc3c9c30496180.html --lua-filter "C:\RLibraries\R4.2.0\rmarkdown\rmarkdown\lua\pagebreak.lua" --lua-filter "C:\RLibraries\R4.2.0\rmarkdown\rmarkdown\lua\latex-div.lua" --embed-resources --standalone --variable bs3=TRUE --section-divs --table-of-contents --toc-depth 3 --variable toc_float=1 --variable toc_selectors=h1,h2,h3 --variable toc_collapsed=1 --variable toc_smooth_scroll=1 --variable toc_print=1 --template "C:\RLibraries\R4.2.0\rmarkdown\rmd\h\default.html" --no-highlight --variable highlightjs=1 --variable theme=journal --mathjax --variable "mathjax-url=https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML" --include-in-header "**C:\Users\fried\AppData\Local\Temp\RtmpGIkvo7\**rmarkdown-str3c9c7840811.html"

**Comment by dcomtois** [comment]
Hi,

From introductory vignette:

Note that as of version 0.9.9, **setting `tmp.img.dir` is no longer required**
**when using `method = "render"`** and can be left to `NA`. It is only 
necessary to define it when a transitory markdown table must be created, as
shown below. Note how narrow the *Graph* column is -- this is actually required, 
since the width of the rendered column is determined by the number of characters
in the cell, rather than the width of the image itself:

```
+---------------+--------|----------------------+---------+
| Variable      | stats  |  Graph               | Valid   |
+===============+========|======================+=========+
| age\          |  ...   | ![](/tmp/ds0001.png) | 978\    |
| [numeric]     |  ...   |                      | (97.8%) |
+---------------+--------+----------------------+---------+
```

---
In short, long paths lead to wide column. There are ways to work around this issue, but they involve lua filters or are quite complicated to implement. If someone wants to give it a shot, you're most welcome to!

**Reply by SimpleAddress4390** [reply]
See comment?

**Reply by dcomtois** [reply]
What do you mean?

**Reply by SimpleAddress4390** [reply]
Well, that is Odd.  I had pasted a 'comment' showing an example of the behavior.  I realized it should have been a 'reply' so I was trying to say, please see the 'comment'.  Now I don't see the comment at all.  I will try this one more time...  ;)

**Reply by SimpleAddress4390** [reply]
Here is a reproducible code example:  There are two code chunk, one with tmp and one without.  AT the end I show the results of each.  What am I missing?  Based on the posts, I shouldn't need the tmp option to get the graphic images of the frequencies, but...

---
title: "dfSummaryTest51.RMD - Confirming tmp usage"
date: "`r Sys.Date()`"
output: 
  html_document: 
    keep_md: no
    theme: spacelab
    toc: yes
    toc_float: true
editor_options: 
  chunk_output_type: inline
    
---

# tmp Directory

I have been unable to use dfSummary without identifying a tmp folder.

Recent posts suggest this should not be necessary.  This RMD is meant to test the process.

```{r setup, echo=FALSE}
suppressWarnings(suppressMessages(library(knitr)))
suppressWarnings(suppressMessages(library(summarytools)))
```

```{r summary tools options 1, eval=TRUE}
summarytools::st_options(
	bootstrap.css     = FALSE,       
	plain.ascii       = FALSE,       
	style             = "rmarkdown", 
	dfSummary.silent  = TRUE,        
	footnote          = NA,          
	descr.silent      = TRUE,        
	tmp.img.dir       = "/tmp",
	subtitle.emphasis = FALSE)       
```

```{r summary tools options 2, eval=FALSE}
summarytools::st_options(
	bootstrap.css     = FALSE,       
	plain.ascii       = FALSE,       
	style             = "rmarkdown", 
	dfSummary.silent  = TRUE,        
	footnote          = NA,          
	descr.silent      = TRUE,        
	subtitle.emphasis = FALSE)       
```

```{r data}

a <- mtcars[,1:3]
b <- PlantGrowth
c <- Indometh[,1:3]

allFiles <- c("a", "b", "c")

```

```{r}
#st_css(main = TRUE, global = TRUE)
```

```{r echo = FALSE, results = 'asis'}

for (df in allFiles) {

  cat("\n### ", df)
  
  print(dfSummary(get(df), 
            headings     = FALSE,
            style        = 'grid',
            graph.magnif = 0.75,
            varnumbers   = FALSE,
            valid.col    = FALSE,
            method = 'render'))
  
}
```
<a href="#top">Back to top</a>


![image](https://github.com/dcomtois/summarytools/assets/7174635/933e48c1-f58e-460f-9684-93072a593616)

![image](https://github.com/dcomtois/summarytools/assets/7174635/97688570-0b2a-4fee-92b7-3931e0c9810a)

**Reply by SimpleAddress4390** [reply]
Let's see if the code works this time.  The example above does not show all the code.

[dfSummaryTest51.RMD.txt](https://github.com/dcomtois/summarytools/files/13439342/dfSummaryTest51.RMD.txt)

**Comment by dcomtois** [comment]
I see now. Indeed it is necessary to specify it. Can't recall exactly why that happens the way it does, but I'm making a note to clarify this in the doc. Thanks for pointing it out!

**Reply by SimpleAddress4390** [reply]
No problem.  It is strange though that the files in tmp somehow get pulled into the folder rmarkdown used to create the final document.  If there was a way to write them there directly, there wouldn't be a need to specify a TMP folder (or clean up manually).    Although I guess these two processed don't really know about each other?

I put a pause in the RMarkdown and can see what is happening:
A folder called something like    C:\Users\fried\AppData\Local\Temp\RtmpE9zrUL   is created and inside of it appears the PNG files (with different names) that exist in the TMP folder:

![2023-11-24 113525 RtmpENaouG](https://github.com/dcomtois/summarytools/assets/7174635/c5de8f97-568a-4b60-b112-7aae530f1403)
![2023-11-24 113551 tmp](https://github.com/dcomtois/summarytools/assets/7174635/66619887-a2fc-440c-a596-166e5bec0dbb)

**Comment by dcomtois** [comment]
You're right that the two processes are not aware of each other. But there
might well be something I overlooked here. I'll look into it.
